pkg/sources/git/git.go

```diff
@@ -472,7 +475,71 @@ func TryAdditionalBaseRefs(repo *git.Repository, base string) (*plumbing.Hash, e
 	return nil, fmt.Errorf("no base refs succeeded for base: %q", base)
 }
 
-// PrepareRepo clones a repo if possible and returns the cloned repo string.
+// PrepareRepoSinceCommit clones a repo starting at the given commitHash and returns the cloned repo path.
+func PrepareRepoSinceCommit(uriString, commitHash string) (string, bool, error) {
+	if commitHash == "" {
+		return PrepareRepo(uriString)
+	}
+	// TODO: refactor with PrepareRepo to remove duplicated logic
+
+	// The git CLI doesn't have an option to shallow clone starting at a commit
+	// hash, but it does have an option to shallow clone since a timestamp. If
+	// the uriString is github.com, then we query the API for the timestamp of the
+	// hash and use that to clone.
+
+	uri, err := url.Parse(uriString)
+	if err != nil {
+		return "", false, fmt.Errorf("unable to parse Git URI: %s", err)
+	}
+
+	if uri.Scheme == "file" || uri.Host != "github.com" {
+		return PrepareRepo(uriString)
+	}
+
+	owner, repoName, found := strings.Cut(uri.Path, "/")
+	if !found {
+		return PrepareRepo(uriString)
+	}
+
+	client := github.NewClient(nil)
+	commit, _, err := client.Git.GetCommit(context.Background(), owner, repoName, commitHash)
+	if err != nil {
+		return PrepareRepo(uriString)
+	}
+	var ts string
```

Comment:
nit: `ts` could be changed to something longer like `timestamp` so it's easier to understand what it is.